chat-bots/src/Data/Chat/Bot.hs

```diff
@@ -83,6 +87,42 @@ newtype Bot m s i o = Bot {runBot :: s -> i -> ListT m (o, s)}
     (Functor, Applicative, Monad, MonadState s, MonadReader i, MonadIO)
     via StateT s (ReaderT i (ListT m))
 
+instance Monad m => Trifunctor.Semigroupal (->) (,) (,) (,) (,) (Bot m) where
+  combine :: (Bot m s i o, Bot m t i' o') -> Bot m (s, t) (i, i') (o, o')
+  combine (Bot b1, Bot b2) = Bot $ \(s, t) (i, i') ->
+    liftA2 (curry (\((o, s'), (o', t')) -> ((o, o'), (s', t')))) (b1 s i) (b2 t i')
+
+instance Functor m => Trifunctor.Semigroupal (->) (,) Either Either (,) (Bot m) where
+  combine :: (Bot m s i o, Bot m t i' o') -> Bot m (s, t) (Either i i') (Either o o')
+  combine (Bot m1, Bot m2) = Bot $ \(s, t) -> \case
+    Left i -> (bimap Left (,t) <$> m1 s i)
+    Right i' -> bimap Right (s,) <$> m2 t i'
+
+instance Monad m => Trifunctor.Semigroupal (->) (,) These These (,) (Bot m) where
+  combine :: (Bot m s i o, Bot m t i' o') -> Bot m (s, t) (These i i') (These o o')
+  combine (Bot b1, Bot b2) = Bot $ \(s, t) -> \case
+    This i -> bimap This (,t) <$> b1 s i
+    That i' -> bimap That (s,) <$> b2 t i'
+    These i i' -> do
+      alignListT (b1 s i) (b2 t i') <&> \case
```

Comment:
Would `liftA2` perform the interleaving in the same way as using `alignListT` here? @masaeedu 

I ask because the definition via `liftA2` is simpler.

Reply:
Unfortunately the `Alignable` instance for `ListT` does a different thing from its `Applicative` instance, so while you can type tetris an answer with `liftA2` I don't think it will do the right thing.

The `Alignable` instance interleaves the effects associated with each of the given `ListT`s, whereas the `Applicative` instance places all the effects of the second list after all the effects of the first.

Reply:
Oh in that case do we want a Semialign instance for ListT?

Reply:
Yes, that would be good.

Reply:
Great. I've got `Semialign` and the 3 relevant monoidal functor instances implemented locally. I'll push them up to this branch once I decide if I want to define monoidal functors in terms of the standard haskell classes or vice versa.

Reply:
Okay i've updated the PR with the new instances. Applicative/Alternative/Semialign are derived from Monoidal.